apps/mark-scan/pat-device-input/src/pin.rs

```diff
@@ -113,7 +103,57 @@ impl Pin {
     }
 }
 
-impl fmt::Display for Pin {
+impl GpioPin {
+    // Exports the pin to be globally accessible from userspace. If the pin is already
+    // exported, does nothing.
+    fn safe_export(&self) -> io::Result<()> {
+        log!(EventId::ConnectToGpioPinInit, "Exporting pin {}", self);
+        if let Err(error) = self.export() {
+            // When the pin is already exported we expect io::ErrorKind::ResourceBusy.
+            // ResourceBusy is unstable so we unconditionally attempt unexport + rexport.
+            // https://github.com/rust-lang/rust/issues/86442
+            log!(
+                EventId::Info,
+                "Error when exporting pin {}: {}. Attempting to unexport and re-export.",
+                self,
+                error
+            );
+
+            self.unexport()?;
+            self.export()?;
+        }
+
+        log!(
+            event_id: EventId::ConnectToGpioPinComplete,
+            disposition: vx_logging::Disposition::Success,
+            message: format!("Successfully exported pin {}", self)
+        );
+        Ok(())
+    }
+
+    // Exports the pin to be globally accessible from userspace. Returns an error if the
+    // pin has already been exported.
+    fn export(&self) -> io::Result<()> {
+        fs::write(EXPORT_PIN_FILEPATH, self.to_string())?;
+        // Without this delay subsequent pin operations may fail with a permission error
+        sleep(EXPORT_DELAY);
+        Ok(())
+    }
+
+    // Unexports the pin, removing access from userspace.
+    fn unexport(&self) -> io::Result<()> {
+        fs::write(UNEXPORT_PIN_FILEPATH, self.to_string())
+    }
+
+    // Sets the pin direction to "in" so its value is readable. Must be called after
+    // `export`
+    fn set_direction_in(&self) -> io::Result<()> {
+        let filepath = format!("/sys/class/gpio/gpio{self}/direction");
```

Comment:
It's not obvious that the `self.fmt(…)` result should be interpolated into the file path without knowing that its implementation just uses `self.address`.

```suggestion
        let filepath = format!("/sys/class/gpio/gpio{}/direction", self.address);
```